### PR TITLE
[REVIEW] Port NVStrings substrings ops to cudf column

### PR DESCRIPTION
**Pull request by davidwendt**
This is porting the NVStrings `slice` and `get` methods to strings_column_view functions.

The `slice` method is a substring function on each string with integer position values.
The `get` can be coded using `slice` where the parameters are `stop=start+1`

The result is a new column strings.

NVStrings methods addressed in this PR:
```
slice()
slice_from()
get()
```

These will be called `substring` and `substring_from` to keep them from getting confused with the cudf `slice` column methods.

### CODE REVIEW

File: cpp/src/strings/substring.cu [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+
+namespace
+{
+
+/**
+ * @brief Used as template parameter to divide size calculation from
+ * the actual string operation within a function.
+ * Useful when most of the logic is identical for both passes.
+ */
+enum TwoPass
```
Comment (harrism) [inline cpp/src/strings/substring.cu]: If there are only two options, why not just a clearly named bool template parameter?

File: cpp/src/strings/substring.cu [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+ * @brief Function logic for the substring API.
+ * This will perform a substring operation on each string
+ * using the provided start, stop, and step parameters.
+ */
+template <TwoPass Pass=SizeOnly>
+struct substring_fn
+{
+    const cudf::column_device_view d_column;
+    int32_t start, stop, step;
+    const int32_t* d_offsets{};
```
Comment (harrism) [inline cpp/src/strings/substring.cu]: Where are these initialized to something other than `nullptr`?

File: cpp/src/strings/substring.cu [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+    // copy the null mask
+    rmm::device_buffer null_mask;
+    cudf::size_type null_count = d_column.null_count();
+    if( d_column.nullable() )
+        null_mask = rmm::device_buffer( d_column.null_mask(),
+                                        cudf::bitmask_allocation_size_bytes(strings_count),
+                                        stream, mr);
+    // build offsets column
+    auto offsets_transformer_itr = thrust::make_transform_iterator( thrust::make_counting_iterator<int32_t>(0),
+        substring_fn<SizeOnly>{d_column, start, stop, step} );
```
Comment (harrism) [inline cpp/src/strings/substring.cu]: The functor is created here without setting its internal pointer members, so they are nullptr. Where do they get set?

File: cpp/src/strings/substring.cu [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+std::unique_ptr<cudf::column> substring( const strings_column_view strings,
+                                         size_type start, size_type stop, size_type step,
+                                         rmm::mr::device_memory_resource* mr,
+                                         cudaStream_t stream )
+{
+    size_type strings_count = strings.size();
+    if( strings_count == 0 )
+        return detail::make_empty_strings_column(mr,stream);
+    CUDF_EXPECTS( start >= 0, "Parameter start must be zero or positive integer.");
+    if( step == 0 )
```
Comment (harrism) [inline cpp/src/strings/substring.cu]: Why not let this be handled by the CUDF_EXPECTS below?

File: cpp/src/strings/substring.cu (hunk 3 above)
Comment (davidwendt) [reply cpp/src/strings/substring.cu]: They are not needed on `SizeOnly` pass just on the `ExecuteOp` pass about 13 lines below this one. So the `SizeOnly` one looks like this:
```
substring_fn<SizeOnly>{d_column, start, stop, step}
```
And the `ExecuteOp` one just below in the same function looks like this:
```
substring_fn<ExecuteOp>{d_column, start, stop, step, d_new_offsets, d_chars};
```
The `d_offsets` and `d_chars` members are not used during `SizeOnly` pass.

File: cpp/src/strings/substring.cu (hunk 1 above)
Comment (davidwendt) [reply cpp/src/strings/substring.cu]: I like seeing `SizeOnly` and `ExecuteOp` in the code inline. I think it makes the code clearer to see these versus knowing that `!SizeOnly` is actually execution or that `!ExecuteOp` means doing a size-only pass.

File: cpp/src/strings/substring.cu [hunk 5]
```diff
@@ -84,19 +84,19 @@ namespace cudf
...
+namespace detail
+{
 
 //
-std::unique_ptr<cudf::column> substring( const strings_column_view strings,
-                                         size_type start, size_type stop, size_type step,
-                                         rmm::mr::device_memory_resource* mr,
-                                         cudaStream_t stream )
+std::unique_ptr<cudf::column> slice_strings( cudf::strings_column_view const& strings,
+                                             cudf::size_type start, cudf::size_type stop=-1, cudf::size_type step=1,
```
Comment (harrism) [inline cpp/src/strings/substring.cu]: You don't need the `cudf::` inside the `cudf` namespace.

File: cpp/src/strings/utilities.cu [hunk 6]
```diff
@@ -145,20 +145,42 @@ std::unique_ptr<column> make_empty_strings_column( rmm::mr::device_memory_resour
...
-{
-    std::lock_guard<std::mutex> guard(g_flags_table_mutex);
-    if( !d_character_flags_table )
+private:
+
+    character_flags_table_type* d_character_flags_table{};
+
+    character_flags_singleton()
     {
         RMM_TRY(RMM_ALLOC(&d_character_flags_table,sizeof(g_character_codepoint_flags),0));
```
Comment (harrism) [inline cpp/src/strings/utilities.cu]: Is this a different singleton than in the other PR? Let's avoid singletons.

Comment (davidwendt) [reply cpp/src/strings/utilities.cu]: This is from a different PR #3159  and should be fixed there now.

**Comment by harrism** [comment]
rerun tests

**Comment by davidwendt** [comment]
rerun tests

**Comment by harrism** [comment]
Sorry @davidwent, I seem to have broken tests/CMakeLists.txt in resolving conflicts.

**Pull request merged**